operations/configuration_management/Write_Ansible_Code_for_CFS.md

```diff
@@ -11,22 +11,55 @@ Help with Ansible can be found in the external Ansible documentation:
+* Avoid using the same CFS configuration/playbook for all nodes and relying on the `hosts` keyword to determine what tasks will run against each node.  Ansible will skip sections of the playbook that have a `hosts` target that doesn't match any nodes in the current inventory/limit, but when multiple types of nodes are configured at the same time with the same configuration, they may end up in the same batch and Ansible run.  This would mean that Ansible has to run through the sections for both types of nodes, taking more time than if the nodes were in separate batches and could skip past the unneeded code.
+
+### Other Tips
+
 * Use loops rather than individual tasks where modules are called multiple times. Some Ansible modules will optimize the command, such as grouping package installations into a single transaction \(Refer to the external [Ansible playbook loops](https://docs.ansible.com/ansible/latest/user_guide/playbooks_loops.html) documentation\).
+* Use Ansible retries for small, recoverable failures.  CFS supports retries on a large scale, but it takes far more time for CFS to detect a failed component and spin up a new session than it does for Ansible to retry a task.
+* Don't use Ansible retries for failures than take a long time to recover from.  Retrying for a significant amount of time on one node can hold up all the other successful nodes in a batch.  If you can't recover from a failure quickly, let the node fail and CFS will separate it out from the successful nodes when new sessions are started.
+* Avoid `any_errors_fatal`.  In addition to not working with all Ansible strategies, this can cause an Ansible run to exit early, and the nodes that didn't have the error will have to start from the beginning of the playbook in the next session.
+* Design playbooks to be run with the `free` Ansible strategy.  This means avoiding situations where all nodes in a batch need to complete a task before moving onto the next, and can save time by allowing nodes to proceed through a playbook at their own pace.
+
+## Ansible Limitations with CFS
+
+Because CFS splits components into multiple batches, and components may also configure at
+different times when they are rebooted, some keywords meant for coordinating the runs of multiple nodes may not work as expected.
+
+* any_errors_fatal - This keyword is intended to stop execution as soon as any node reports a failure.  However, this will only stop execution for the current batch.
+* run_once - This keyword is intended to limit a task to running on a single node. However this will only cause the task to be run once per batch.
+* serial - This keyword is intended to be used to limit runs to a small number of nodes at a time, such as during a rolling upgrade.  However, this will only function within the batch, so more nodes may be running the task than intended when multiple batches are running
```

Comment:
```suggestion
| `serial` | This keyword is intended to limit runs to a small number of nodes at a time, such as during a rolling upgrade. However, this will only function within the batch, so more nodes may be running the task than intended when multiple batches are running. |
```